File: src/main/java/com/bajdcc/util/HashListMapEx.java
package com.bajdcc.util;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 哈希表
 * <p>
 * T-&gt;id,V=list,id= list indexof T
 * </p>
 *
 * @author bajdcc
 */
public class HashListMapEx<K, V> {

	public Map<K, V> map = new HashMap<>();

	public List<V> list = new ArrayList<>();

	public boolean contains(K k) {
		return map.containsKey(k);
	}

	public void add(K k, V v) {
		V vv = map.put(k, v);
		if (vv == null) {
			list.add(v);
		} else {
			list.set(list.indexOf(vv), v);
		}
	}

	public boolean put(K k, V v) {
		V vv = map.put(k, v);
		if (vv == null) {
			list.add(v);
			return false;
		} else {
			list.set(list.indexOf(vv), v);
			return true;
		}
	}

	public V get(K k) {
		return map.get(k);
	}

	public int indexOf(V v) {
		return list.indexOf(v);
	}
}

File: src/main/java/com/bajdcc/util/HashListMapEx.kt
package com.bajdcc.util

import com.google.common.collect.HashBiMap

/**
 * 双向哈希表
 * [K] -> [V]
 * @author bajdcc
 */
class HashListMapEx<K, V> {

    private var map = HashBiMap.create<K, V>()

    operator fun contains(k: K) = map.containsKey(k)

    fun add(k: K, v: V) = map.put(k, v)

    fun put(k: K, v: V) = map.put(k, v) != null

    operator fun get(k: K): V = map[k]!!

    fun copy() = HashListMapEx<K, V>().also {
        it.map = this.map
    }

    fun toList() : List<V> = map.toList().map { it.second }
}